### ISSUE TITLE
no-double-slash-css-comments

### ISSUE THREAD

**Issue by kangax** [stylelint/stylelint#661]
Double slash (`//...`) comments are not supported by CSS (afaik, they're ignored but for a different reason than being a comment). They could also lead to some issues during minification (if minifier is not aware of them) and/or with [urls that start with double slash](https://discourse.wicg.io/t/single-line-comments/1048).

[More details here](http://stackoverflow.com/a/20192639/130652).

A rule to prevent `//...` comments and use proper `/*...*/` instead seems to make sense.

**Comment by jeddy3** [comment]
Great suggestion for a rule!

It's an interesting one though as the linter supports scss syntax via the `postcss-scss` parser. So we'll need to make sure the name of this rule correctly represents it's behaviour across both parsers.

I think if the rule is called `comment-no-double-slash` then it should check for both `//...` and `// ...` comments. If you're using the standard css parser you'll get a lint warning on `//...` and a parse error on `// ...`. If you're using the scss parser you'll get a lint warning for both. Both groups will probably have different reasons for enabling the rule, but the behaviour of the rule itself is consistent.

How does that sound?

**Comment by jeddy3** [comment]
Actually, this might be a little more problematic than I first thought as [it seems](https://github.com/postcss/postcss/issues/362) postcss doesn't parse `//...` either.

Ideas?

**Comment by kangax** [comment]
Interesting. However, I don't get a parse error on either `//...` or `// ...` (and I'm using CSS; unless SCSS parser kicks in somehow without me explicitly enabling it).

**Comment by kangax** [comment]
I just did a simple test (via a custom stylelint plugin):

```js
root.walkDecls(node => {
  if (node.prop.indexOf('//') === 0) {
    report(...)
  }
});
```

..and it caught both "//position:absolute" (property == "//position") and "// position: absolute" (property == "//").

**Comment by jeddy3** [comment]
Ooooh, I see. You're specifically referring to the `//...` commenting of declarations. Yep, the standard parser can cope with those. It also appears fine [with other double slash hacks](http://www.xanthir.com/b4U10):

```css
// I'm a double slash comment outside of rule with two braces at the end - postcss parses me as a selector {}

a {
 // I'm a double slash comment in a rule with a colon at the end - postcss parses me a prop:
}
```

The following trips up the standard parser though:

```css
// comment without the two braces at the end

a {
 // comment without a colon at the end
}
```

So, perhaps this rule should be called `no-double-slash-comment-hacks` because, as you pointed out at the beginning, they aren't really comments and they are more attached to the _sheet_ itself. That will also simplify the implementation as we'll just check for props and selectors that start with double slash. We can then keep the way clear for a possible `comment-no-double-slash` rule in the future that checks only scss syntax, where `//...` is, I believe, actually parsed as a comment node. How does that sound?

> and I'm using CSS; unless SCSS parser kicks in somehow without me explicitly enabling it).

FYI, it needs to be explicitly enabled.

**Comment by kangax** [comment]
Hm, can't say I'm a big fan of that name (no-double-slash-comment-hacks). I think it might be confusing to people — why "hacks" and not regular "//" comments (most don't know that those aren't parsed as comments)? Maybe "no-double-slash-css-comments" then?

**Comment by jeddy3** [comment]
> why "hacks" and not regular "//" comments (most don't know that those aren't parsed as comments)?

Good point.

> Maybe "no-double-slash-css-comments" then

Sounds good to me :)  

@davidtheclark  Any thoughts on any of this, or shall I mark it as ready to go?

**Comment by davidtheclark** [comment]
Sounds good.

**Comment by jeddy3** [comment]
@kangax This is ready to go :)

I'm not sure if you'll find it useful, but these two existing rules might be provide good pointers:

* [selector-root-no-composition](https://github.com/stylelint/stylelint/blob/master/src/rules/selector-root-no-composition/index.js) - checking for `// comment {}` selector comments.
* [declaration-no-important](https://github.com/stylelint/stylelint/blob/master/src/rules/declaration-no-important/index.js) - checking for for `// comment:  ` declaration comments.

**Comment by kangax** [comment]
https://github.com/stylelint/stylelint/pull/675 — let me know if I got everything right!

**Comment by jeddy3** [comment]
Done in https://github.com/stylelint/stylelint/pull/675
